lib/mix/lib/mix/dep/adder.ex

```diff
@@ -0,0 +1,132 @@
+    new_line = "{:#{app}, #{dep_opts}}"
+
+    formatted = format_to_string(mix_exs)
+
+    if mix_exs == formatted do
+      lines = String.split(mix_exs, "\n")
+      check_for_app!(lines, app)
+
+      case add_dep_line(lines, [], new_line) do
+        new_lines when is_list(new_lines) ->
+          {:ok, new_line, new_lines |> Enum.join("\n") |> format_to_string()}
+
+        :deps_not_found ->
+          {:error, new_line, :deps_not_found}
+      end
+    else
+      {:error, new_line, :not_formatted}
+    end
+  end
+
+  defp format_to_string(str),
+    do: str |> Code.format_string!() |> IO.iodata_to_binary() |> Kernel.<>("\n")
+
+  defp get_runtime_string(opts) do
+    if Keyword.has_key?(opts, :runtime) and !opts[:runtime] do
+      ", runtime: false"
+    else
+      ""
+    end
+  end
+
+  defp get_only_string(opts) do
+    case opts[:only] do
+      [] -> ""
+      [env] -> ", only: :#{env}"
+      envs -> ", only: [" <> (envs |> Enum.map(&(":" <> &1)) |> Enum.join(", ")) <> "]"
+    end
+  end
+
+  defp get_version_str(opts, app) do
+    cond do
+      path = opts[:path] ->
+        ~s|path: "#{path}"|
+
+      github = opts[:github] ->
+        ~s|github: "#{github}"| <> ref_or_tag(opts)
+
+      version = opts[:version] || latest_hex_version(app) ->
+        case version do
+          "~> " <> _ = version -> ~s|"#{version}"|
+          ">= " <> _ = version -> ~s|"#{version}"|
+          "0.0.0" -> ~s|">= 0.0.0"|
+          version -> ~s|"~> #{version}"|
+        end
+    end
+  end
+
+  defp ref_or_tag(opts) do
+    cond do
+      ref = opts[:ref] -> ~s|, ref: "#{ref}"|
+      tag = opts[:tag] -> ~s|, tag: "#{tag}"|
+      true -> ""
+    end
+  end
+
+  defp check_for_app!(lines, app) do
+    lines
+    |> Enum.find(fn line ->
+      line
+      |> String.trim()
+      |> String.starts_with?("{:" <> app)
+    end)
+    |> case do
+      nil ->
+        :ok
+
+      found ->
+        Mix.raise("#{app} already exists in mix.exs as #{String.trim(found)}")
+    end
+  end
+
+  defp latest_hex_version(_app) do
+    # TODO: Get latest version from Hex
+    "0.0.0"
+  end
+
+  defp add_dep_line([line1, line2 | rest], acc, dep_line) do
+    cond do
+      String.trim(line1) == "defp deps do" && String.trim(line2) == "[" ->
+        dep_line =
+          if additional_deps?(rest) do
+            dep_line <> ","
+          else
+            dep_line
+          end
+
+        Enum.reverse(acc) ++ [line1, line2, indent(dep_line, line2) | rest]
+
+      String.trim(line1) == "defp deps do" && String.trim(line2) == "[]" ->
+        Enum.reverse(acc) ++
+          [line1, indent("[", line1), indent(dep_line, line2), indent("]", line1) | rest]
+
+      String.trim(line1) == "defp deps do" && String.starts_with?(String.trim(line2), "[{:") ->
+        Enum.reverse(acc) ++
+          [line1, String.replace(line2, ~r/\[/, "[#{dep_line}, ", global: false) | rest]
+
+      true ->
+        add_dep_line([line2 | rest], [line1 | acc], dep_line)
+    end
```

Comment:
Same as `check_for_app!/2` … When parsing `AST` we few options:
1. Invalid `mix.exs` file (typo, missing file etc.)
2. if `:deps` option is `variable` or raw `List` - if so do: `[{:my_dep, …} | variable]` or simply insert `:my_dep` as first element of `List`
3. if `:deps` option call 0-arity function - in such case for last expression try point 2
4. otherwise display proper message

Also in current code such example would fail:
```elixir
defp deps, do: […]
```

Reply:
Similar to above, this would currently result in:
```
Could not find `defp deps do` in your mix.exs. Please add this line to mix.exs manually:
{:foo, ">= 0.0.0"}
```
although tweaking the code to be able to detect that one-liner would be pretty trivial.